File: algorithms/cpp/replaceElementswithGreatestElementonRightSide/replaceElementswithGreatestElementonRightSide.cpp
// Source: https://leetcode.com/problems/replace-elements-with-greatest-element-on-right-side/
// Author: Miao Zhang
// Date:   2021-04-21

class Solution {
public:
    vector<int> replaceElements(vector<int>& arr) {
        int g = -1;
        for (int i = arr.size() - 1; i >= 0; i--) {
            g = max(g, exchange(arr[i], g));
        }
        return arr;
    }
};

File: algorithms/python/replaceElementswithGreatestElementonRightSide/replaceElementswithGreatestElementonRightSide.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/replace-elements-with-greatest-element-on-right-side/
# Author: Miao Zhang
# Date:   2021-04-21

class Solution:
    def replaceElements(self, arr: List[int]) -> List[int]:
        res = [0 for _ in range(len(arr))]
        res[-1] = -1
        for i in range(len(arr) - 2, -1, -1):
            res[i] = max(res[i + 1], arr[i + 1])
        return res
